python312Packages.zarr: 2.18.7 -> 3.1.0

bcdarwin opened the pull request:
The following packages need to be updated upstream to the Zarr 3.x. API:

 - [x] `python312Packages.tifffile`: https://github.com/cgohlke/tifffile/issues/282
 - [x] `python312Packages.multiscale-spatial-image`: https://github.com/spatial-image/multiscale-spatial-image/pull/108
 - [ ] `python312Packages.ome-zarr`: https://github.com/ome/ome-zarr-py/issues/463
 - [x] `python312Packages.reproject`

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bcdarwin committed:
python312Packages.zarr: 2.18.7 -> 3.0.2

dotlambda commented on pkgs/development/python-modules/zarr/default.nix:
```diff
@@ -5,44 +5,67 @@
+  version = "3.0.0";
+  pyproject = true;
 
   disabled = pythonOlder "3.8";
```
```suggestion
  disabled = pythonOlder "3.11";
```

bcdarwin replied:
Done.

dotlambda commented on pkgs/development/python-modules/zarr/default.nix:
```diff
@@ -5,44 +5,67 @@
   pname = "zarr";
-  version = "2.18.3";
-  format = "pyproject";
+  version = "3.0.0";
```
```suggestion
  version = "3.0.2";
```

dotlambda commented:
> * [ ]  python312Packages.multiscale-spatial-image

I can't find an upstream issue for that. Mind opening one?

doronbehar committed:
python312Packages.zarr: 3.0.2 -> 3.0.4

Changelog: https://github.com/zarr-developers/zarr-python/releases/tag/v3.0.4

doronbehar committed:
python312Packages.zarr: don't test flaky tests

This issue might still be observed though:

https://github.com/zarr-developers/zarr-python/issues/2732

doronbehar commented:
> > * [ ]   python312Packages.multiscale-spatial-image
> 
> I can't find an upstream issue for that. Mind opening one?

Done, also for tifffile, and updated the top comment of the PR. Also, I pushed a few more commits related to the main topic of the PR. I also added there `reproject` which is another direct dependencies of `zarr`.

doronbehar commented:
After working a bit with the patch here and with zarr v3 directly, I noticed that many features of it are missing in comparison to version 2.x. Even if the dependencies mentioned in the PR's top comment will adapt to v3, I think we should still have a `zarr2` attribute and perform a gradual transition. Do you think @dotlambda maybe we should maybe at the moment create a `zarr3` (and `zarr2`) attribute(s) and later on handle the dependencies? In a similar fashion to what we did with Numpy.. It still be nice to let Nixpkgs users be able to use `zarr3` if they want to without maintaining a fork. I anticipate that the dependent packages will take time to adapt anyway, also possibly due to missing features.

doronbehar commented:
Due to:

- https://github.com/NixOS/nixpkgs/pull/396376

And due to the fact non of the upstream issues in dependent projects have had any progress, I decided to decouple the `numcodecs` update done here, to a separate PR:

- https://github.com/NixOS/nixpkgs/pull/396492

I also rebased this PR to fix the natural merge conflict due to #396376 .

doronbehar committed:
python312Packages.numcodecs: 0.15.1 -> 0.16.0

doronbehar committed:
python312Packages.numcodecs: 0.16.0 -> 0.16.1

doronbehar committed:
python313Packages.zarr: 3.0.4 -> 3.1.0

Changelog: https://github.com/zarr-developers/zarr-python/releases/tag/v3.1.0

doronbehar committed:
python312Packages.tifffile: no with lib; in meta

doronbehar committed:
python313Packages.tifffile: 2025.5.10 -> 2025.6.11

Changelog: https://github.com/cgohlke/tifffile/blob/v2025.6.11/CHANGES.rst

doronbehar commented:
A few updates: 

- We have now zarr 3.1.0
- Only 2 dependent packages are waiting for upstream to support zarr 3.x - top comment of PR was updated accordingly.
- I added `numcodecs_0_15` and `zarr_2` derivations, that should be useful if someone insists not to update to zarr 3.x, by using an overlay - please give me feedback on the idea if you can.
- Along with the above, I'm considering to mark as broken the 2 dependent packages that don't support yet zarr 3.x, depending on the zarr version, and hence people will be able to build them with the same overlay that downgrades both `numcodecs_0_15` and `zarr_2`. What do you think?

bcdarwin commented:
It looks like these PRs are active and will likely be merged before the next NixOS release cycle, so I'd be fine with merging.

If we do proceed, I'm not sure how much effort/mostly-duplicated code we should put into supporting Zarr 2 stuff in tree though.

bcdarwin commented:
https://github.com/spatial-image/multiscale-spatial-image/pull/108 was merged and it looks like they'll make a release soon as well so we can clean this PR up a bit.

For now I've just rebased onto master.

bcdarwin committed:
python313Packages.multiscale-spatial-image: 2.0.2 -> 2.0.3

bcdarwin committed:
python313Packages.xarray-dataclass: init at 3.0.0

bcdarwin committed:
python313Packages.spatial-image: 1.2.1 -> 1.2.3

doronbehar commented:
I think we can wait to the weekend for https://github.com/ome/ome-zarr-py/issues/463 too - it won't be a big deal a few more days after 7 months this PR was open :).

bcdarwin committed:
python313Packages.ome-zarr: 0.11.1 -> 0.12rc1

bcdarwin commented:
Updated to the ome-zarr release candidate and dropped the pinned versions of zarr and numcodecs.

bcdarwin committed:
python313Packages.anndata: 0.11.4 -> 0.12.2

bcdarwin committed:
python313Packages.legacy-api-wrap: init at 1.4.1

bcdarwin commented:
> ## `nixpkgs-review` result

Looks like `spectral-cube` and `pvextract` are already broken on master; `halide` seems to have failed with an illegal instruction on my machine (and the build might be flaky on master as well?); and the `devShellTools` test seems to have timed out on my underpowered build?

doronbehar commented:
I managed to build locally halide with this PR (rebased upon latest `master`). I admit I haven't checked `tests.devShellTools.nixos` which is sourced in `nixpkgs/nixos/tests/docker-tools-nix-shell.nix` - because I hardly believe this PR will have an effect on it.

Hence, I'm merging :tada:.
